airbyte-integrations/bases/base-python/base_python/integration.py

```diff
@@ -87,34 +79,94 @@ def spec(self, logger: AirbyteLogger) -> ConnectorSpecification:
         raw_spec = pkgutil.get_data(self.__class__.__module__.split(".")[0], "spec.json")
         return ConnectorSpecification.parse_obj(json.loads(raw_spec))
 
-    def check(self, logger: AirbyteLogger, config: json) -> AirbyteConnectionStatus:
+    @abstractmethod
+    def check(self, logger: AirbyteLogger, config: Mapping[str, any]) -> AirbyteConnectionStatus:
         """
         Tests if the input configuration can be used to successfully connect to the integration e.g: if a provided Stripe API token can be used to connect
         to the Stripe API.
         """
         raise Exception("Not Implemented")
 
+
+class Source(Integration):
+    # can be overridden to change an input state
+    def read_state(self, state_path: str) -> Dict[str, any]:
+        if state_path:
+            state_obj = json.loads(open(state_path, "r").read())
+        else:
+            state_obj = {}
+        state = defaultdict(dict, state_obj)
+        return state
+
     def discover(self, logger: AirbyteLogger, config: json) -> AirbyteCatalog:
         """
         Returns an AirbyteCatalog representing the available streams and fields in this integration. For example, given valid credentials to a
         Postgres database, returns an Airbyte catalog where each postgres table is a stream, and each table column is a field.
         """
         raise Exception("Not Implemented")
 
-
-class Source(Integration):
-    def __init__(self):
-        super().__init__()
-
     def read(
-        self, logger: AirbyteLogger, config: json, catalog: ConfiguredAirbyteCatalog, state_path: Dict[str, any]
+            self, logger: AirbyteLogger, config: json, catalog: ConfiguredAirbyteCatalog, state_path: Dict[str, any]
     ) -> Generator[AirbyteMessage, None, None]:
         """
         Returns a generator of the AirbyteMessages generated by reading the source with the given configuration, catalog, and state.
         """
         raise Exception("Not Implemented")
 
 
-class Destination(Integration):
-    def __init__(self):
-        super().__init__()
+class Destination(Integration, ABC):
+    logger = AirbyteLogger()
+
+    @abstractmethod
+    def write(self, logger: AirbyteLogger, config: Mapping[str, any], configured_catalog: ConfiguredAirbyteCatalog) -> Iterable[AirbyteMessage]:
```

Comment:
I assume this is unfinished since I would expect write to consume and not produce.

BTW, I wonder if the highest level Destination interface, instead of being `write` (including in java) should instead be an `onMessage`

Reply:
With 80% certainty I think we should always have a very barebones abstraction that contains `write` (like this one) because that is the protocol-level operation. We can have deeper abstractions but the user should always have the option of taking complete ownership over the `write` operation. 

The return type is incorrect -- it was anticipating some of the destination checkpointing stuff but I think for now I'm just going to remove it.